File: mobile/ios/packages/town-crier-domain/Sources/Entities/AuthSession.swift
import Foundation

/// Represents an authenticated user session with tokens and profile.
public struct AuthSession: Equatable, Sendable {
  public let accessToken: String
  public let idToken: String
  public let expiresAt: Date
  public let userProfile: UserProfile
  /// The user's subscription tier extracted from the JWT access token.
  /// Defaults to `.free` when the claim is absent.
  public let subscriptionTier: SubscriptionTier

  public init(
    accessToken: String,
    idToken: String,
    expiresAt: Date,
    userProfile: UserProfile,
    subscriptionTier: SubscriptionTier = .free
  ) {
    self.accessToken = accessToken
    self.idToken = idToken
    self.expiresAt = expiresAt
    self.userProfile = userProfile
    self.subscriptionTier = subscriptionTier
  }

  /// Whether the session's access token has expired.
  public var isExpired: Bool {
    expiresAt <= Date.now
  }
}

File: mobile/android/domain/src/main/kotlin/uk/towncrierapp/domain/auth/AuthSession.kt
package uk.towncrierapp.domain.auth

import uk.towncrierapp.domain.subscriptions.SubscriptionTier
import java.time.Clock
import java.time.Instant

/** An authenticated user session: tokens, identity, and the tier read off the access token's JWT claim. */
public data class AuthSession(
    val accessToken: String,
    val idToken: String,
    val expiresAt: Instant,
    val userProfile: UserProfile,
    val subscriptionTier: SubscriptionTier = SubscriptionTier.FREE,
) {
    /** Whether the access token has expired as of [clock] (injected — domain code never reads the wall clock itself). */
    public fun isExpired(clock: Clock): Boolean = !expiresAt.isAfter(clock.instant())
}
